[quant] Add _FusedModule type to capture all fused modules for quantization

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47415 [quant] Remove nn.quantized.ReLU module and nn.quantized.functional.relu
* **#47484 [quant] Add _FusedModule type to capture all fused modules for quantization**

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

jerryzh168 committed:
[quant] Add _FusedModule type to capture all fused modules for quantization

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]

z-a-f commented on torch/nn/intrinsic/modules/fused.py:
```diff
@@ -1,52 +1,56 @@
-class ConvReLU1d(torch.nn.Sequential):
-    r"""This is a sequential container which calls the Conv 1d and ReLU modules.
+# Used for identifying intrinsic modules used in quantization
+class _FusedModule(torch.nn.Sequential):
+    pass
```
question: is having a common ancestor the right pattern? So far we were creating maps and list for that purpose

jerryzh168 replied:
yeah, i feel common parent is better than the list, what do you think?

z-a-f replied:
I don't have strong feelings either way: Having a common pattern feels better to me, but the lists/maps is the way we used to do

jerryzh168 replied:
yeah I have removed the list/maps in this PR, we can use isinstance(model, _FusedModule) to check for this in the future.

z-a-f commented on torch/nn/qat/modules/linear.py:
```diff
@@ -2,7 +2,7 @@
 from torch.nn.intrinsic import LinearReLU
 
-class Linear(nn.Linear):
+class Linear(nn.Linear, _FusedModule):
```
I feel like patterns such as this might cause a diamond problem if the first class is also inheriting from sequential

jerryzh168 replied:
right, but we probably won't have that I think

z-a-f reviewed (approved):
lgtm

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add _FusedModule type to capture all fused modules for quantization"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D24774703](https://our.internmc.facebook.com/intern/diff/D24774703)

[ghstack-poisoned]
